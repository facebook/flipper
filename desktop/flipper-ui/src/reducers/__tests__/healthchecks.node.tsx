/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {
  default as reducer,
  startHealthchecks,
  finishHealthchecks,
  updateHealthcheckResult,
  acknowledgeProblems,
} from '../healthchecks';
import type {FlipperDoctor} from 'flipper-common';

const HEALTHCHECKS: FlipperDoctor.Healthchecks = {
  ios: {
    label: 'iOS',
    isSkipped: false,
    isRequired: true,
    healthchecks: [
      {
        key: 'ios.sdk',
        label: 'SDK Installed',
        run: async (_env: FlipperDoctor.EnvironmentInfo) => {
          return {
            hasProblem: false,
            message: ['ios.sdk--installed', {platforms: ['ios']}],
          };
        },
      },
    ],
  },
  android: {
    label: 'Android',
    isSkipped: false,
    isRequired: true,
    healthchecks: [
      {
        key: 'android.sdk',
        label: 'SDK Installed',
        run: async (_env: FlipperDoctor.EnvironmentInfo) => {
          return {
            hasProblem: true,
            message: ['android.sdk--no_ANDROID_HOME'],
          };
        },
      },
    ],
  },
  common: {
    label: 'Common',
    isSkipped: false,
    isRequired: false,
    healthchecks: [
      {
        key: 'common.openssl',
        label: 'OpenSSL Istalled',
        run: async (_env: FlipperDoctor.EnvironmentInfo) => {
          return {
            hasProblem: false,
            message: ['common.openssl--installed', {output: 'not found'}],
          };
        },
      },
    ],
  },
};

test('startHealthCheck', () => {
  const res = reducer(undefined, startHealthchecks(HEALTHCHECKS));
  expect(res).toMatchSnapshot();
});

test('updateHealthcheckResult', () => {
  let res = reducer(undefined, startHealthchecks(HEALTHCHECKS));
  res = reducer(
    res,
    updateHealthcheckResult('android', 'android.sdk', {
      message: ['android.sdk--installed', {output: 'some error message'}],
      isAcknowledged: false,
      status: 'SUCCESS',
    }),
  );
  expect(res).toMatchSnapshot();
});

test('finish', () => {
  let res = reducer(undefined, startHealthchecks(HEALTHCHECKS));
  res = reducer(
    res,
    updateHealthcheckResult('ios', 'ios.sdk', {
      message: ['ios.sdk--installed', {platforms: ['ios']}],
      isAcknowledged: false,
      status: 'SUCCESS',
    }),
  );
  res = reducer(
    res,
    updateHealthcheckResult('android', 'android.sdk', {
      message: ['android.sdk--installed', {output: 'success'}],
      isAcknowledged: false,
      status: 'SUCCESS',
    }),
  );
  res = reducer(
    res,
    updateHealthcheckResult('common', 'common.openssl', {
      message: ['common.openssl--installed', {output: 'not found'}],
      isAcknowledged: false,
      status: 'SUCCESS',
    }),
  );
  res = reducer(res, finishHealthchecks());
  expect(res).toMatchSnapshot();
});

test('statuses updated after healthchecks finished', () => {
  let res = reducer(undefined, startHealthchecks(HEALTHCHECKS));
  res = reducer(
    res,
    updateHealthcheckResult('android', 'android.sdk', {
      message: ['android.sdk--not_installed', {output: 'some error message'}],
      isAcknowledged: false,
      status: 'FAILED',
    }),
  );
  res = reducer(
    res,
    updateHealthcheckResult('ios', 'ios.sdk', {
      message: ['ios.sdk--installed', {platforms: ['ios']}],
      isAcknowledged: false,
      status: 'SUCCESS',
    }),
  );
  res = reducer(
    res,
    updateHealthcheckResult('common', 'common.openssl', {
      message: ['common.openssl--installed', {output: 'found'}],
      isAcknowledged: false,
      status: 'SUCCESS',
    }),
  );
  res = reducer(res, finishHealthchecks());
  expect(res).toMatchSnapshot();
});

test('acknowledgeProblems', () => {
  let res = reducer(undefined, startHealthchecks(HEALTHCHECKS));
  res = reducer(
    res,
    updateHealthcheckResult('ios', 'ios.sdk', {
      isAcknowledged: false,
      status: 'FAILED',
    }),
  );
  res = reducer(
    res,
    updateHealthcheckResult('android', 'android.sdk', {
      isAcknowledged: false,
      status: 'SUCCESS',
    }),
  );
  res = reducer(
    res,
    updateHealthcheckResult('common', 'common.openssl', {
      isAcknowledged: false,
      status: 'FAILED',
    }),
  );
  res = reducer(res, finishHealthchecks());
  res = reducer(res, acknowledgeProblems());
  expect(res).toMatchSnapshot();
});
