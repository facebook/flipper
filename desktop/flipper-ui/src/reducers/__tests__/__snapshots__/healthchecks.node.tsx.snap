// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`acknowledgeProblems 1`] = `
{
  "acknowledgedProblems": [
    "ios.sdk",
    "common.openssl",
  ],
  "healthcheckReport": {
    "categories": {
      "android": {
        "checks": {
          "android.sdk": {
            "key": "android.sdk",
            "label": "SDK Installed",
            "result": {
              "isAcknowledged": true,
              "status": "SUCCESS",
            },
          },
        },
        "key": "android",
        "label": "Android",
        "result": {
          "isAcknowledged": true,
          "status": "SUCCESS",
        },
      },
      "common": {
        "checks": {
          "common.openssl": {
            "key": "common.openssl",
            "label": "OpenSSL Istalled",
            "result": {
              "isAcknowledged": true,
              "status": "FAILED",
            },
          },
        },
        "key": "common",
        "label": "Common",
        "result": {
          "isAcknowledged": true,
          "status": "FAILED",
        },
      },
      "ios": {
        "checks": {
          "ios.sdk": {
            "key": "ios.sdk",
            "label": "SDK Installed",
            "result": {
              "isAcknowledged": true,
              "status": "FAILED",
            },
          },
        },
        "key": "ios",
        "label": "iOS",
        "result": {
          "isAcknowledged": true,
          "status": "FAILED",
        },
      },
    },
    "result": {
      "isAcknowledged": true,
      "status": "FAILED",
    },
  },
}
`;

exports[`finish 1`] = `
{
  "acknowledgedProblems": [],
  "healthcheckReport": {
    "categories": {
      "android": {
        "checks": {
          "android.sdk": {
            "key": "android.sdk",
            "label": "SDK Installed",
            "result": {
              "isAcknowledged": false,
              "message": [
                "android.sdk--installed",
                {
                  "output": "success",
                },
              ],
              "status": "SUCCESS",
            },
          },
        },
        "key": "android",
        "label": "Android",
        "result": {
          "isAcknowledged": false,
          "status": "SUCCESS",
        },
      },
      "common": {
        "checks": {
          "common.openssl": {
            "key": "common.openssl",
            "label": "OpenSSL Istalled",
            "result": {
              "isAcknowledged": false,
              "message": [
                "common.openssl--installed",
                {
                  "output": "not found",
                },
              ],
              "status": "SUCCESS",
            },
          },
        },
        "key": "common",
        "label": "Common",
        "result": {
          "isAcknowledged": false,
          "status": "SUCCESS",
        },
      },
      "ios": {
        "checks": {
          "ios.sdk": {
            "key": "ios.sdk",
            "label": "SDK Installed",
            "result": {
              "isAcknowledged": false,
              "message": [
                "ios.sdk--installed",
                {
                  "platforms": [
                    "ios",
                  ],
                },
              ],
              "status": "SUCCESS",
            },
          },
        },
        "key": "ios",
        "label": "iOS",
        "result": {
          "isAcknowledged": false,
          "status": "SUCCESS",
        },
      },
    },
    "result": {
      "status": "SUCCESS",
    },
  },
}
`;

exports[`startHealthCheck 1`] = `
{
  "acknowledgedProblems": [],
  "healthcheckReport": {
    "categories": {
      "android": {
        "checks": {
          "android.sdk": {
            "key": "android.sdk",
            "label": "SDK Installed",
            "result": {
              "status": "IN_PROGRESS",
            },
          },
        },
        "key": "android",
        "label": "Android",
        "result": {
          "status": "IN_PROGRESS",
        },
      },
      "common": {
        "checks": {
          "common.openssl": {
            "key": "common.openssl",
            "label": "OpenSSL Istalled",
            "result": {
              "status": "IN_PROGRESS",
            },
          },
        },
        "key": "common",
        "label": "Common",
        "result": {
          "status": "IN_PROGRESS",
        },
      },
      "ios": {
        "checks": {
          "ios.sdk": {
            "key": "ios.sdk",
            "label": "SDK Installed",
            "result": {
              "status": "IN_PROGRESS",
            },
          },
        },
        "key": "ios",
        "label": "iOS",
        "result": {
          "status": "IN_PROGRESS",
        },
      },
    },
    "result": {
      "status": "IN_PROGRESS",
    },
  },
}
`;

exports[`statuses updated after healthchecks finished 1`] = `
{
  "acknowledgedProblems": [],
  "healthcheckReport": {
    "categories": {
      "android": {
        "checks": {
          "android.sdk": {
            "key": "android.sdk",
            "label": "SDK Installed",
            "result": {
              "isAcknowledged": false,
              "message": [
                "android.sdk--not_installed",
                {
                  "output": "some error message",
                },
              ],
              "status": "FAILED",
            },
          },
        },
        "key": "android",
        "label": "Android",
        "result": {
          "isAcknowledged": false,
          "status": "FAILED",
        },
      },
      "common": {
        "checks": {
          "common.openssl": {
            "key": "common.openssl",
            "label": "OpenSSL Istalled",
            "result": {
              "isAcknowledged": false,
              "message": [
                "common.openssl--installed",
                {
                  "output": "found",
                },
              ],
              "status": "SUCCESS",
            },
          },
        },
        "key": "common",
        "label": "Common",
        "result": {
          "status": "SUCCESS",
        },
      },
      "ios": {
        "checks": {
          "ios.sdk": {
            "key": "ios.sdk",
            "label": "SDK Installed",
            "result": {
              "isAcknowledged": false,
              "message": [
                "ios.sdk--installed",
                {
                  "platforms": [
                    "ios",
                  ],
                },
              ],
              "status": "SUCCESS",
            },
          },
        },
        "key": "ios",
        "label": "iOS",
        "result": {
          "status": "SUCCESS",
        },
      },
    },
    "result": {
      "isAcknowledged": false,
      "status": "FAILED",
    },
  },
}
`;

exports[`updateHealthcheckResult 1`] = `
{
  "acknowledgedProblems": [],
  "healthcheckReport": {
    "categories": {
      "android": {
        "checks": {
          "android.sdk": {
            "key": "android.sdk",
            "label": "SDK Installed",
            "result": {
              "isAcknowledged": false,
              "message": [
                "android.sdk--installed",
                {
                  "output": "some error message",
                },
              ],
              "status": "SUCCESS",
            },
          },
        },
        "key": "android",
        "label": "Android",
        "result": {
          "status": "IN_PROGRESS",
        },
      },
      "common": {
        "checks": {
          "common.openssl": {
            "key": "common.openssl",
            "label": "OpenSSL Istalled",
            "result": {
              "status": "IN_PROGRESS",
            },
          },
        },
        "key": "common",
        "label": "Common",
        "result": {
          "status": "IN_PROGRESS",
        },
      },
      "ios": {
        "checks": {
          "ios.sdk": {
            "key": "ios.sdk",
            "label": "SDK Installed",
            "result": {
              "status": "IN_PROGRESS",
            },
          },
        },
        "key": "ios",
        "label": "iOS",
        "result": {
          "status": "IN_PROGRESS",
        },
      },
    },
    "result": {
      "status": "IN_PROGRESS",
    },
  },
}
`;
